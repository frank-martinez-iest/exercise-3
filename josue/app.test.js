import { fireEvent, getByText, getById, getAllByText, getNodeText, getByLabelText, getByPlaceholderText} from '@testing-library/dom';
import '@testing-library/jest-dom/extend-expect';
import { JSDOM } from 'jsdom';
import fs  from 'fs';
import path  from 'path';
import { async } from 'regenerator-runtime';

const html = fs.readFileSync(path.resolve(__dirname,'./index.html'),'utf-8');

let dom
let window
let document
let container
let scriptElement

describe('Unit test login feature',()=>{
    beforeEach(()=> {
        const scriptContent = fs.readFileSync(path.resolve(__dirname,'./js/app-test.js'),'utf-8');
        dom = new JSDOM(html, {runScripts :'dangerously'});
        window = dom.window;
        document = window.document;
        container = dom.window.document.body;
        scriptElement = document.createElement('script');
        scriptElement.textContent = scriptContent;
        document.head.appendChild(scriptElement);
    });
    it('should save to localStorage',()=>{
        const dummyUser = {"name":"Juan Lopez","phone":"8332041352","email":"juanperez@gmail.com","password":"12345","password_confirmation":"12345"}
        localStorage.setItem('8332041352', dummyUser);
        expect(localStorage.setItem).toHaveBeenLastCalledWith('8332041352',dummyUser);
        console.log('++localstorage: ',localStorage)
    });

    it('renders Sign In form',()=>{
        expect(container.querySelector('#form')).toBeInTheDocument();
    });
    it('returns error message when user submits an empty form',async () => {
       const loginBtn = getByText(container, 'Sign In');
       await fireEvent.click(loginBtn);

       const errorMessageElement = await container.querySelector('.error-message');
       expect(errorMessageElement.innerText).not.toBe('')
    });
    it('must not contain register fields',()=>{
        /* REGISTER FIELDS DIV*/
        expect(container.querySelector('#register-fields')).not.toBeInTheDocument();
    });
    it('updates submit btn when register btn gets clicked', async()=>{
        const registerBtn = getByText(container,'Register');
        fireEvent.click(registerBtn);
        
        expect(container.querySelector('#submit').innerText).toBe('Register');
    });
    it('adds register fields when register btn gets clicked',async()=>{
        const registerBtn = getByText(container,'Register');
        fireEvent.click(registerBtn);
        /* REGISTER FIELDS DIV*/
        expect(container.querySelector('#register-fields')).not.toBeNull();
        /* Email Field*/
        expect(getByText(container, 'Email')).toBeInTheDocument();
        /* Password Field*/
        expect(getByText(container,'Password')).toBeInTheDocument();
        /* Password confirmation Field*/
        expect(getByText(container, 'Confirm Password')).toBeInTheDocument();
    });
    it('shows error message when register fields are empty',async()=>{
        const registerBtn = getByText(container,'Register');
        fireEvent.click(registerBtn);
        await fireEvent.click(registerBtn);

       const errorMessageElement = await container.querySelector('.error-message');
       expect(errorMessageElement.innerText).not.toBe('')

    });
})