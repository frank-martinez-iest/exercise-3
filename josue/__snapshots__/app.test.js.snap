// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test login feature renders Sign In form 1`] = `
<body>
  

  
  <nav
    class="navbar"
  >
    
    
    <ul
      class="navbar__items"
    >
      
      
      <li
        class="navbar__logo"
      >
        
        
        <a
          href="/"
        >
          
          Bank
          <em>
            App
          </em>
          
        
        </a>
        
      
      </li>
      
      
      <li
        class="navbar__link"
      >
        
        
        <button
          class="btn btn--secondary"
          id="signBtn"
        >
          Log In
        </button>
        
      
      </li>
      
      
      <li
        class="navbar__link"
      >
        
        
        <button
          class="btn btn--primary"
          id="registerBtn"
        >
          Register
        </button>
        
      
      </li>
      
    
    </ul>
    
  
  </nav>
  
  
  <main>
    
    
    <section
      class="account"
    >
      

      
      <div
        class="account__left-wrapper"
      >
        
        
        <form
          class="form"
          id="form"
        >
          
          
          <div
            class="form-group"
          >
            
            
            <label
              for="name"
            >
              
              Name
            
            </label>
            
            
            <input
              autocomplete=""
              id="name"
              name="Name"
              placeholder="Lorem Perez"
              required=""
              type="text"
            />
            
            
            <span
              class="error-message"
            />
            
          
          </div>
          
          
          <div
            class="form-group"
          >
            
            
            <label
              for="phone"
            >
              
              Phone
            
            </label>
            
            
            <input
              autocomplete=""
              id="phone"
              name="Phone"
              placeholder="8334563456"
              required=""
              type="tel"
            />
            
            
            <span
              class="error-message"
            />
            

          
          </div>
          
          
          
          
          <button
            class="btn btn--primary"
            id="submit"
          >
            Sign In
          </button>
          
        
        </form>
        
      
      </div>
      
      
      <div
        class="account__splash-img"
      >
        
        
        <h1>
          Welcome!
        </h1>
        
      
      </div>
      
      
      
    
    </section>
    
  
  </main>
  
  
  <!--This is not read by jest-->
  
  
  <script
    src="js/app.js"
    type="module"
  />
  



</body>
`;
